### PR TITLE
[BISERVER-13264] When using a custom AccessVoter, cut/paste does not …

### PR DESCRIPTION
**Pull request by mchen-len-son**
…check for write permission on destination
- added accessVoter consulting to file moving method
- added unit-test

### CODE REVIEW

**Comment by mchen-len-son** [comment]
@rmansoor @rfellows @pentaho-nbaker @pamval please review

File: extensions/src/org/pentaho/platform/web/http/api/resources/operations/CopyFilesOperation.java [hunk 1]
```diff
@@ -262,7 +263,9 @@ private void copyRenameMode( RepositoryFile repoFile ) {
       repositoryFile =
         getRepository().createFile( destDir.getId(), duplicateFile, data, acl, null );
     }
-
+    if ( repositoryFile == null ) {
+      throw new UnifiedRepositoryAccessDeniedException( "Not enough permission to write file." );
```
Comment (mbatchelor) [inline extensions/src/org/pentaho/platform/web/http/api/resources/operations/CopyFilesOperation.java]: This message needs to be externalized to a properties file as do several other messages in this class.

File: extensions/src/org/pentaho/platform/web/http/api/resources/services/FileService.java [hunk 2]
```diff
@@ -743,7 +745,12 @@ public void doCopyFiles( String pathId, Integer mode, String params ) {
     CopyFilesOperation copyFilesOperation =
       new CopyFilesOperation( getRepository(), getRepoWs(), Arrays.asList( sourceFileIds ), path, mode );
 
-    copyFilesOperation.execute();
+    try {
+      copyFilesOperation.execute();
+    } catch ( UnifiedRepositoryAccessDeniedException e ) {
+      // let's convert it to an appropriate one for REST resource
+      throw new IllegalArgumentException();
```
Comment (mbatchelor) [inline extensions/src/org/pentaho/platform/web/http/api/resources/services/FileService.java]: You should pass on the original exception here - this is swallowing what the actual error was.

File: extensions/src/org/pentaho/platform/web/http/api/resources/services/FileService.java [hunk 3]
```diff
@@ -342,6 +342,8 @@ public void doMoveFiles( String destPathId, String params ) throws FileNotFoundE
       }
     } catch ( IllegalArgumentException e ) {
       throw e;
+    } catch ( UnifiedRepositoryAccessDeniedException e ) {
+      throw new IllegalArgumentException();
```
Comment (mbatchelor) [inline extensions/src/org/pentaho/platform/web/http/api/resources/services/FileService.java]: You should pass on the original exception here - this is swallowing what the actual error was.

File: extensions/src/org/pentaho/platform/web/http/api/resources/FileResource.java [hunk 4]
```diff
@@ -286,6 +286,8 @@ public Response doMove( @PathParam ( "pathId" ) String destPathId, String params
     } catch ( FileNotFoundException e ) {
       logger.error( Messages.getInstance().getErrorString( "FileResource.DESTINATION_PATH_UNKNOWN", destPathId ), e );
       return buildStatusResponse( Response.Status.NOT_FOUND );
+    } catch ( IllegalArgumentException e ) {
+      return buildStatusResponse( Response.Status.FORBIDDEN );
```
Comment (mbatchelor) [inline extensions/src/org/pentaho/platform/web/http/api/resources/FileResource.java]: Should have a logger.error here with the exception.

File: repository/src/org/pentaho/platform/repository2/unified/jcr/JcrRepositoryFileDao.java [hunk 5]
```diff
@@ -846,11 +846,25 @@ private void internalCopyOrMove( final Serializable fileId, final String destRel
...
         RepositoryFile file = getFileById( fileId );
-        RepositoryFileAcl acl = aclDao.getAcl( fileId );
-        if ( !accessVoterManager.hasAccess( file, RepositoryFilePermission.WRITE, acl, PentahoSessionHolder
-            .getSession() ) ) {
-          return null;
+        if ( !copy ) {
+          RepositoryFileAcl acl = aclDao.getAcl( fileId );
+          if ( !accessVoterManager.hasAccess( file, RepositoryFilePermission.WRITE, acl,
+            PentahoSessionHolder.getSession() ) ) {
+            throw new AccessDeniedException( "Not enough permissions to remove file." );
```
Comment (mbatchelor) [inline repository/src/org/pentaho/platform/repository2/unified/jcr/JcrRepositoryFileDao.java]: NLS-ize this message (and any others in this class)

File: extensions/src/org/pentaho/platform/web/http/api/resources/FileResource.java (hunk 4 above)
Comment (mbatchelor) [inline extensions/src/org/pentaho/platform/web/http/api/resources/FileResource.java]: As stated on 8/16, you need to log the error here.

File: extensions/src/org/pentaho/platform/web/http/api/resources/operations/CopyFilesOperation.java (hunk 1 above)
Comment (mbatchelor) [inline extensions/src/org/pentaho/platform/web/http/api/resources/operations/CopyFilesOperation.java]: This message needs to be externalized to a properties file as do several other messages in this class.

File: extensions/src/org/pentaho/platform/web/http/api/resources/services/FileService.java (hunk 3 above)
Comment (mbatchelor) [inline extensions/src/org/pentaho/platform/web/http/api/resources/services/FileService.java]: You should pass on the original exception here - this is swallowing what the actual error was.

File: extensions/src/org/pentaho/platform/web/http/api/resources/services/FileService.java (hunk 2 above)
Comment (mbatchelor) [inline extensions/src/org/pentaho/platform/web/http/api/resources/services/FileService.java]: You should pass on the original exception here - this is swallowing what the actual error was.

File: repository/src/org/pentaho/platform/repository2/unified/jcr/JcrRepositoryFileDao.java (hunk 5 above)
Comment (mbatchelor) [inline repository/src/org/pentaho/platform/repository2/unified/jcr/JcrRepositoryFileDao.java]: NLS-ize this message (and any others in this class)

File: repository/src/org/pentaho/platform/repository2/unified/jcr/JcrRepositoryFileDao.java [hunk 6]
```diff
@@ -846,11 +846,25 @@ private void internalCopyOrMove( final Serializable fileId, final String destRel
...
+          }
+        }
+
+        // check that user has permissions to write to the destination folder
+        RepositoryFile destFolder = getFile( destRelPath );
+        if ( destFolder != null ) {
+          RepositoryFileAcl destFolderAcl = aclDao.getAcl( destFolder.getId() );
+          if ( !accessVoterManager.hasAccess( destFolder, RepositoryFilePermission.WRITE, destFolderAcl,
+            PentahoSessionHolder.getSession() ) ) {
+            throw new AccessDeniedException( "Not enough permissions to write file." );
```
Comment (mbatchelor) [inline repository/src/org/pentaho/platform/repository2/unified/jcr/JcrRepositoryFileDao.java]: NLS-ize this message (and any others in this class)

**Comment by mchen-len-son** [comment]
@mbatchelor sorry for delay, i've updated the PR.
please, take a look.
thanks.

**Comment by pamval** [comment]
@mbatchelor please review this pr again

**Pull request merged**